### PR TITLE
Expand functionality on knowing about retractions

### PR DESCRIPTION
**Pull request by svenvanderburg**
### Added
* `pubkey` option to methods of `NanopubClient` that allows searching for publications 
signed with the given pubkey. For these methods:
    - `find_nanopubs_with_text`
    - `find_nanopubs_with_pattern`
    - `find_things`
* `filter_retracted` option to methods of `NanopubClient` that allows searching for publications 
    that are not retracted. For these methods:
    - `find_nanopubs_with_text`
    - `find_nanopubs_with_pattern`
    - `find_things`
* `NanopubClient.find_retractions_of` method to search retractions of a given nanopublication.
* `Publication.signed_with_public_key` property: the public key that the publication was signed with.
* `Publication.is_test_publication` property: denoting whether this is a publicaion on the test server.

I did not add a `is_retracted` property to the `Publication` class, because we would have to instantiate a `NanopubClient` in the `Publication` object. Then you get all kinds of circular import problems, and the problem that you cannot specify whether you want to use `test` or `prod` server. So users just have to use `NanopubClient.find_retractions_of` to know whether a publication was retracted (I added documentation on this).

### CODE REVIEW

File: nanopub/client.py [hunk 1]
```diff
@@ -62,40 +67,18 @@ def find_nanopubs_with_text(self, text: str, max_num_results: int = 1000):
         """
         if len(text) == 0:
             return []
-
+        endpoint = 'find_signed_nanopubs_with_text'
```
Comment (svenvanderburg) [inline nanopub/client.py:70]: We actually never use the 'normal' find_nanopubs_with_text endpoint anymore.

Comment (raar1) [reply nanopub/client.py:70]: That's fine. I don't think unsigned nanopubs are really useful, given the whole point of the infrastructure being to assign authorship/reputation to assertions.

File: nanopub/client.py [hunk 2]
```diff
@@ -180,44 +142,57 @@ def find_things(self, type: str, searchterm: str = ' ',
...
+        else:
+            public_key = None
+
+        results = self.find_nanopubs_with_pattern(pred=namespaces.NPX.retracts,
+                                                  obj=rdflib.URIRef(uri),
+                                                  pubkey=public_key,
+                                                  filter_retracted=False)
+        return [result['np'] for result in results]
 
     def _query_grlc(self, params, endpoint):
```
Comment (raar1) [inline nanopub/client.py:197]: I'd say this essentially solves #20 (as much as it will ever be solved)

File: nanopub/publication.py [hunk 3]
```diff
@@ -28,10 +29,11 @@ class Publication:
         source_uri (str): The URI of the nanopublication that this Publication represents (if
             applicable)
         introduces_concept: The concept that is introduced by this Publication.
-
+        signed_with_public_key: The public key that this Publication is signed with.
+        is_test_publication: Whether this is a test publication
     """
 
-    def __init__(self, rdf=None, source_uri=None):
+    def __init__(self, rdf, source_uri=None):
```
Comment (raar1) [inline nanopub/publication.py:36]: Should we give a warning/error if the ```rdf``` passed is e.g. ```None```?

**Review by raar1 — APPROVED** [review]
Great PR. Documentation is good, and extensive tests. I've left a couple of comments/questions.

File: nanopub/publication.py (hunk 3 above)
Comment (svenvanderburg) [reply nanopub/publication.py]: This now gives an error if you don't specify it, which is already a huge step forward. 

Checking that the `rdf` is not `None` would be nice, we could also check that it is an instance of `rdflib.Graph`. We don't really do much type & value checking anywhere throughout the library, but we could. Maybe you should create an issue if you want to do this (but I'm fine with not doing it)

I've added type hinting so it is clear what to pass which I think is enough in this case. We don't expect the user to use this actually.

**Pull request merged**